Split YAML installing topic

snneji opened the pull request:
Fixes #3361

- I split the original installing topic into 4 pages: Prereqs, Serving, Eventing, Optional extensions
- Added a redirect for the old file name (any-kubernetes-cluster.md) to the index for the installing section
- Updated links to the old topic throughout the docs
- Added a Next Steps section at the bottom of the installing topics with cross references to the Optional extensions topic and the topic to install the other component (in case they want to install Eventing and Serving).

snneji commented:
@omerbensaadon I have a few questions:
1. I decided to add a Next Steps section with cross references instead of adding a next button. This is because:
    1. Buttons at the bottom of the page are not used anywhere else on this site. Users might miss it. 
    2. If there is a next button, users who don't want to install Eventing after Serving would be taken through the Eventing page.
2. Would it be a good idea to add a link for getting started with using Knative Serving/Eventing to the Next Steps sections? 
3. Should we mention to install the Knative CLI at some point in the installation?
4. Is my choice of where to put the redirect for the page I deleted OK (it goes to the index page for installing)?

omerbensaadon commented:
@snneji I trust whatever solution you come up with as a docs writer, these are recommendations from the UX WG and we can discuss more and test with users once they have been delivered!

RichardJJG commented:
/lgtm

snneji commented:
/assign @vagababov

omerbensaadon commented:
//TODO on Deliveery Set up an issue that standardizes linking to the Pre-Requisite page, as discussed with @abrennan89  in #3404

abrennan89 commented on docs/eventing/broker/kafka-broker.md:
````diff
@@ -263,7 +263,7 @@ data:
 ```
 
-To change the logging level to `DEBUG`, you need to: 
+To change the logging level to `DEBUG`, you need to:
````
```suggestion
To change the logging level to `DEBUG`, you must:
```

abrennan89 commented on docs/eventing/samples/helloworld/helloworld-go/README.md:
```diff
@@ -246,8 +246,10 @@ folder) you're ready to build and deploy the sample app.
 
-      _Note_: you can also use injection based on labels with the
-      [Eventing Sugar Controller](../../../../install/any-kubernetes-cluster.md).
+      **Note:** you can also use injection based on labels with the
+      Eventing Sugar Controller.
+      For how to install the Eventing Sugar Controller, see
```
Small nit, I'd make "sugar controller" lowercase, but we don't have standards yet for things like this afaik
cc @omerbensaadon

snneji replied:
There has been a suggestion to have a word and phrase list. The docs are a bit inconsistent on capitalisation and I think having a list would make these decisions easier.

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+showlandingtoc: "false"
+---
+
+This topic describes how to install Knative Eventing.
+You can install the Knative Eventing component by applying YAML files using the `kubectl` CLI.
```
```suggestion
This topic describes how to install Knative Eventing by applying YAML files using the `kubectl` CLI.
```

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+This topic describes how to install Knative Eventing.
+You can install the Knative Eventing component by applying YAML files using the `kubectl` CLI.
+
+When installing Knative, you can install the Serving component, Eventing component, or both on your cluster.
+For how to install the Serving component, see [Installing Serving using YAML files](./install-serving-with-yaml.md)
```
I'd remove this since this section is specific to eventing

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+See [Knative Prerequisites](./prerequisites.md).
+
+
+## Install the Eventing component
```
Installing? Can we standardize on using gerunds in headings maybe?

snneji replied:
I'm used to using imperatives in headings. I can't find a rule in the Google style guide and the Knative docs are not consistent on this. I usually revert to what I'm used to if I cannot find a rule. 

Is there a forum where we can discuss style decisions such as these so that we can all be doing the same thing?

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+
+## Install the Eventing component
+
+To install the Eventing component:
```
seems unnecessary to have this line

snneji replied:
This text will be expanded as part of issue #3378

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+
+To install the Eventing component:
+
+1. Install the required custom resources:
```
```suggestion
1. Install the required custom resource definitions (CRDs):
```

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
````diff
@@ -0,0 +1,225 @@
+```
+
+
+## Optional: Install a default Channel (messaging) layer
+
````
I don't think it's correctly worded here to say default, I would assume this means the channel supplied with a default install, i.e. InMemoryChannel even though I know you probably mean "install a channel and set it as the default type". Things like Kafka have additional requirements afaik like a Kafka cluster and installing additional CRDs. Maybe have someone from @knative/eventing-wg-leads check this part.

snneji replied:
This wording was already on the page and is in the v0.21 docs (step 3 [here](https://knative.dev/docs/install/any-kubernetes-cluster/#installing-the-eventing-component)).

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+
+## Optional: Install a default Channel (messaging) layer
+
+The tabs below expand to show instructions for installing a default Channel (messaging) layer.
```
I'd lowercase channel, also let's not use parenthesis where we can avoid it maybe? I don't know that we need to say messaging after channel every time.

snneji replied:
Just did a find and replace for all the uppercase occurrences of channel. It seems pretty consistently uppercase in this topic but it makes more sense to me for it to be lowercase as you suggest. The same goes for broker. I'll update them all.

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+## Optional: Install a default Channel (messaging) layer
+
+The tabs below expand to show instructions for installing a default Channel (messaging) layer.
+Follow the procedure for the Channel of your choice:
```
again, I'd just either lowercase "channel" unless we're referring to a specific Channel object

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+{{< /tabs >}}
+
+
+## Optional: Install a Broker (Eventing) layer:
```
Again, remove the parenthesis, (Eventing) doesn't add value here IMO

abrennan89 commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+
+After installing Knative Eventing:
+
+- If you want to add extra features to your installation, see [Installing optional extensions](./install-extensions.md).
```
I'd make all the channels and broker stuff that aren't part of core eventing optional steps too instead of including them above

snneji replied:
As in putting them on the optional topic? The way I imagined it was a serving topic and eventing topic and and and optional extensions topic. Are the broker and channels stuff optional extensions? Or do you think I should change the topic name from "Installing optional extensions" to something else?

snneji replied:
If you are recommending splitting the optional topic into "optional for serving" and "optional for eventing", I am concerned about creating clutter in the subnav. Splitting further will mean that we will end up with four topics (five including the Prerequisites) where there was only one before. 

If we agree that splitting the optional topic is a good idea, I think we should reorganise the subnav.

abrennan89 commented on docs/install/install-extensions.md:
```diff
@@ -0,0 +1,278 @@
+showlandingtoc: "false"
+---
+
+To add extra features to your Knative Serving or Eventing installation, you can install extensions
+by applying YAML files using the `kubectl` CLI.
+
+For information about the YAML files in the Knative Serving release, see
+[Installation files](./installation-files.md).
+
+
+# Prerequisites
+
+Before you install any optional extensions, you must install Knative Serving or Eventing.
+See [Installing Serving using YAML files](./install-serving-with-yaml.md)
+and [Installing Eventing using YAML files](./install/install-eventing-with-yaml.md).
```
I'd split this into optional stuff for serving and optional stuff for eventing, so that they can have their own prerequisites etc. We only point to the install files for serving above which is confusing.

abrennan89 commented on docs/install/install-extensions.md:
````diff
@@ -0,0 +1,278 @@
+kubectl apply -f {{< artifact org="knative-sandbox" repo="eventing-couchdb" file="couchdb.yaml" >}}
+```
+
+To learn more about the Apache CouchDB source, read the [documentation](https://github.com/knative-sandbox/eventing-couchdb/blob/main/source/README.md).
````
see above re rewording for links

snneji replied:
see my above comment

abrennan89 commented on docs/install/install-extensions.md:
````diff
@@ -0,0 +1,278 @@
+
+{{% tab name="VMware Sources and Bindings" %}}
+
+The following command installs the VMware Sources and Bindings:
+
+```bash
+kubectl apply -f {{< artifact org="vmware-tanzu" repo="sources-for-knative" file="release.yaml" >}}
+```
+
+To learn more about the VMware sources and bindings, try
+[our samples](https://github.com/vmware-tanzu/sources-for-knative/tree/master/samples/README.md).
+
````
be consistent with capitalizing, this is why I'd recommend everything lower case, e.g. sources, brokers, bindings, channels, etc.

snneji replied:
see my above comment about fixing style issues

abrennan89 commented on docs/install/install-serving-with-yaml.md:
```diff
@@ -0,0 +1,380 @@
+showlandingtoc: "false"
+---
+
+This topic describes how to install Knative Serving.
+You can install the Knative Serving component by applying YAML files using the `kubectl` CLI.
```
```suggestion
This topic describes how to install Knative Serving by applying YAML files using the `kubectl` CLI.
```

abrennan89 commented on docs/install/install-serving-with-yaml.md:
```diff
@@ -0,0 +1,380 @@
+This topic describes how to install Knative Serving.
+You can install the Knative Serving component by applying YAML files using the `kubectl` CLI.
+
+When installing Knative, you can install the Serving component, Eventing component, or both on your cluster.
+For how to install the Eventing component, see [Installing Eventing using YAML files](./install-eventing-with-yaml.md)
```
again, I'd remove this

abrennan89 commented on docs/install/install-serving-with-yaml.md:
```diff
@@ -0,0 +1,380 @@
+
+## Install the Serving component
+
+To install the serving component:
```
I'd remove this line - I think a lot of the same comments from above apply for serving too

abrennan89 commented on docs/install/install-serving-with-yaml.md:
```diff
@@ -0,0 +1,380 @@
+    {{< /tab >}} {{< /tabs >}}
+
+
+## Next steps
+
+After installing Knative Serving:
+
+- If you want to add extra features to your installation, see [Installing optional extensions](./install-extensions.md).
+- If you want to install the Knative Eventing component, see [Installing Eventing using YAML files](./install-eventing-with-yaml.md)
```
as next steps for both I'd also recommend installing `kn`

abrennan89 commented on docs/serving/autoscaling/README.md:
```diff
@@ -11,7 +11,9 @@ For more information about which metrics can be used to control the Autoscaler,
 ## Optional autoscaling configuration tasks
 
-* Configure your Knative deployment to use the Kubernetes [Horizontal Pod Autoscaler (HPA)](../../install/any-kubernetes-cluster.md#optional-serving-extensions) instead of the default KPA.
+* Configure your Knative deployment to use the Kubernetes Horizontal Pod Autoscaler (HPA)
+instead of the default KPA.
+For how to install HPA, see [Install optional Eventing extensions](../../install/install-extensions.md#install-optional-serving-extensions).
```
It's odd to me to add an additional line for this rather than how it was formatted before, I think it's less readable.
It might be worth checking out "minimalism in tech writing" as a concept if you haven't come across this before, a quick google search will give a lot of links on the topic.

snneji replied:
I've used the standard cross referencing format from the Google style guide. I am aware of minimalism, but, when following a link, readers should have a reasonable expectation of what they will find at the target.

Readers need to be aware that, after following this link, they will find info about _installing_ HPA and not general information about HPA.

abrennan89 replied:
I disagree, but not strongly enough to block a PR on it 🙂

abrennan89 commented on docs/serving/autoscaling/autoscaling-concepts.md:
```diff
@@ -11,7 +11,9 @@ This section covers conceptual information about which Autoscaler types are supp
 Knative Serving supports the implementation of Knative Pod Autoscaler (KPA) and Kubernetes' Horizontal Pod Autoscaler (HPA). The features and limitations of each of these Autoscalers are listed below.
 
-**IMPORTANT:** If you want to use Kubernetes Horizontal Pod Autoscaler (HPA), you must install it after you install [Knative Serving](../../install/any-kubernetes-cluster.md#optional-serving-extensions).
+**IMPORTANT:** If you want to use Kubernetes Horizontal Pod Autoscaler (HPA),
+you must install it after you install Knative Serving.
+For how to install HPA, see [Install optional Eventing extensions](../../install/install-extensions.md#install-optional-serving-extensions).
```
same comment as above, I'd prefer we use minimalism principles and don't add extra lines that don't add value

snneji replied:
see my above comment about cross ref format

abrennan89 commented on docs/serving/autoscaling/autoscaling-concepts.md:
```diff
@@ -21,7 +23,7 @@ Knative Serving supports the implementation of Knative Pod Autoscaler (KPA) and
 ### Horizontal Pod Autoscaler (HPA)
 
-* Not part of the Knative Serving core, and must be enabled after [Knative Serving installation](../../install/any-kubernetes-cluster.md#optional-serving-extensions).
+* Not part of the Knative Serving core, and must be enabled after.
```
Removing this makes the sentence incomplete, enabled after what...?

abrennan89 commented on docs/serving/using-auto-tls.md:
```diff
@@ -29,26 +29,26 @@ Knative supports the following Auto TLS modes:
 1.  Using HTTP-01 challenge
 
-    - In this type, your cluster does not need to be able to talk to your DNS server. You just 
+    - In this type, your cluster does not need to be able to talk to your DNS server. You just
     need to map your domain to the IP of the cluser ingress.
```
```suggestion
    - In this type, your cluster does not need to be able to talk to your DNS server. You must map your domain to the IP of the cluser ingress.
```

abrennan89 commented on docs/serving/using-auto-tls.md:
```diff
@@ -29,26 +29,26 @@ Knative supports the following Auto TLS modes:
 ## Before you begin
 
-You must meet the following prerequisites to enable auto TLS:
+You must meet the following prerequisites to enable Auto TLS:
```
@omerbensaadon I'd like a standard way to refer to things like this - just a note for if you're looking for doc issues

omerbensaadon replied:
Started adding things here: https://docs.google.com/spreadsheets/d/1p1_kBUd6ZvonxHkMcEJPayf6QIpExuFf5cFq0ptar7I/edit#gid=0

abrennan89 replied:
Nice, thanks. There's actually an open issue for this here https://github.com/knative/docs/issues/170 if you want to take that one?
I think we should use guidance from the Kubernetes style guide https://kubernetes.io/docs/contribute/style/style-guide/ and the Google developer's style guide where possible https://developers.google.com/style/word-list?hl=en

abrennan89 reviewed (changes requested):
Left some comments.

I think it's worth looking into minimalism in technical writing from a UX perspective, cc @omerbensaadon

snneji commented on docs/install/install-eventing-with-yaml.md:
```diff
@@ -0,0 +1,225 @@
+
+## Prerequisites
+
+Before installation, you must meet the prerequisites.
+See [Knative Prerequisites](./prerequisites.md).
+
```
We discussed this in the Docs WG meeting. Our compromise was to have the Prereqs in another topic and provide a link.

omerbensaadon replied:
The idea being that we wouldn't have to update things everywhere. 

In the future, we could have this appear as a "Toggle" block which references a single source of truth for Pre-Reqs...

abrennan89 replied:
Another reason I hate markdown - no content re-use / inclusion 🙃 Do you know if mkdocs adds any functionality to provide that?
If we used asciidoc we would just have one content module that we `include::` in different places, just sayin' 😂 

Fair enough if we're just linking to it for now, but as a reader I just think it would annoy me having to navigate away to another page to read prereqs and then come back here to do the actual install.

snneji commented on docs/install/install-eventing-with-yaml.md:
````diff
@@ -0,0 +1,225 @@
+```bash
+kubectl get pods --namespace knative-eventing
+```
+
````
This issue was about splitting the YAML installing topic. I can make a new issue to improve the Verify the installation sections in eventing and serving after this PR has been merged.

snneji commented on docs/install/prerequisites.md:
```diff
@@ -0,0 +1,29 @@
+
+Before installing Knative, you must meet the following prerequisites:
+
+## System requirements
+
+For prototyping purposes, Knative will work on most local deployments of Kubernetes.
+For example, you can use a local, one-node cluster that has 2 CPU and 4GB of memory.
+
+For production purposes, it is recommended that:
+- If you have only one node in your cluster, you will need at least 6 CPUs, 6 GB of memory, and 30 GB of disk storage.
+- If you have multiple nodes in your cluster, for each node you will need at least 2 CPUs, 4 GB of memory, and 20 GB of disk storage.
+<!--TODO: Verify these requirements-->
+
+**NOTE:** The system requirements provided are recommendations only.
+The requirements for your installation may vary, depending on whether you use optional components, such as a networking layer.
+
+## Prerequisites
+
+Before installation, you must meet the following prerequisites:
+
+- You have a cluster that uses Kubernetes v1.18 or newer.
+- You have installed the [`kubectl` CLI](https://kubernetes.io/docs/tasks/tools/install-kubectl/).
+- Your Kubernetes cluster must have access to the internet, since Kubernetes needs to be able to fetch images.
```
I copied this directly from the text that is currently on the installing page. Since this PR is only to split the installing topic as decided in the docs WG meeting, would it be OK to open an issue to edit this topic after this PR is merged?

snneji commented on docs/install/install-extensions.md:
```diff
@@ -0,0 +1,278 @@
+{{< /tab >}} {{< /tabs >}}
+
+
+## Install optional Eventing extensions
```
see my comment above.

snneji commented on docs/install/install-extensions.md:
````diff
@@ -0,0 +1,278 @@
+kubectl apply -f {{< artifact org="google" repo="knative-gcp" file="cloud-run-events.yaml" >}}
+```
+
+To learn more about the Cloud Pub/Sub source, try
+[our sample](../eventing/samples/cloud-pubsub-source/README.md).
+
+To learn more about the Cloud Storage source, try
+[our sample](../eventing/samples/cloud-storage-source/README.md).
+
+To learn more about the Cloud Scheduler source, try
+[our sample](../eventing/samples/cloud-scheduler-source/README.md).
+
+To learn more about the Cloud Audit Logs source, try
+[our sample](../eventing/samples/cloud-audit-logs-source/README.md).
````
Since this story was to split existing content and not to edit it, I think fixing style related things like link text should be part of a separate issue.

snneji commented on docs/install/install-serving-with-yaml.md:
````diff
@@ -0,0 +1,380 @@
+   ```
+
+
+## Install a networking layer
````
see comment above about optional docs

snneji commented:
Hi @abrennan89, thanks for reviewing my PR! I applied most of your feedback and replied to your comments.

This PR is to split the existing information on the installing topic as we agreed in the docs WG meeting. I also fixed links because I removed a topic. I tried to avoid writing new content or editing existing content because this was out of scope. 

Because this PR deletes the file `any-kubernetes-cluster.md`, it will become difficult to rebase if anyone makes changes to that topic before this PR is merged. I'm aware that there are issues with capitalisation, link text, and other style issues. It might be better to open new docs issues to fix these problems as they existed before I split the topic. Can we merge this PR and deal with fixing the style later? 

I can open issues to fix the following items from your review:
1. Improve the verify the installation sections. Add example output from the command.
2. Edit each new installing topic for style (incl. capitalisation, link text, etc.)
3. Fix the prereqs topic —  combine headings logically

I left comments unresolved in where I think further discussion might be useful.

abrennan89 reviewed (approved):
/lgtm
/approve

Thanks for all the work on this @snneji 🙂 anything else can be followed up later, I don't see any reason to block this further.
